### PR TITLE
{cae}[foss/2018b] OpenFOAM v1906

### PR DESCRIPTION
**Pull request by mboisson**

### CODE REVIEW

**Comment by mboisson** [comment]
This PR requires https://github.com/easybuilders/easybuild-easyblocks/pull/1772

**Comment by omula** [comment]
@mboisson I have tried it with foss/2018a and it doesn't work for me.

```
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Comm_rank'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Abort'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `ompi_mpi_comm_null'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `ompi_mpi_group_null'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Alltoallv'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Scatterv'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Comm_group'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Allreduce'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Bsend'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Buffer_detach'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `ompi_mpi_byte'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Recv'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Wait'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Test'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Alltoall'
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so: undefined reference to `MPI_Comm_size
```

It seems that mpi compilers are not being used to compile pstream.
```
wmake  mpi (mpi=EASYBUILDMPI)
wmake mpi
    ln: ./lnInclude
Making dependency list for source file PstreamGlobals.C
Making dependency list for source file UPstream.C
Making dependency list for source file UOPwrite.C
Making dependency list for source file UIPread.C
g++ -std=c++11 -m64 -DOPENFOAM=1906 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O2 -ftree-vectorize -march=native -fno-math-errno -std=c++11 -fuse-ld=bfd  -DNoRepository -ftemplate-depth-100   -Wno-old-style-cast -Wno-unused-local-typedefs -Wno-array-bounds -Wno-deprecated-declarations -fpermissive -IlnInclude -I. -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OpenFOAM/lnInclude -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OSspecific/POSIX/lnInclude   -fPIC -c UOPwrite.C -o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UOPwrite.o
g++ -std=c++11 -m64 -DOPENFOAM=1906 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O2 -ftree-vectorize -march=native -fno-math-errno -std=c++11 -fuse-ld=bfd  -DNoRepository -ftemplate-depth-100   -Wno-old-style-cast -Wno-unused-local-typedefs -Wno-array-bounds -Wno-deprecated-declarations -fpermissive -IlnInclude -I. -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OpenFOAM/lnInclude -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OSspecific/POSIX/lnInclude   -fPIC -c UIPread.C -o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UIPread.o
g++ -std=c++11 -m64 -DOPENFOAM=1906 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O2 -ftree-vectorize -march=native -fno-math-errno -std=c++11 -fuse-ld=bfd  -DNoRepository -ftemplate-depth-100   -Wno-old-style-cast -Wno-unused-local-typedefs -Wno-array-bounds -Wno-deprecated-declarations -fpermissive -IlnInclude -I. -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OpenFOAM/lnInclude -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OSspecific/POSIX/lnInclude   -fPIC -c UPstream.C -o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UPstream.o
g++ -std=c++11 -m64 -DOPENFOAM=1906 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O2 -ftree-vectorize -march=native -fno-math-errno -std=c++11 -fuse-ld=bfd  -DNoRepository -ftemplate-depth-100   -Wno-old-style-cast -Wno-unused-local-typedefs -Wno-array-bounds -Wno-deprecated-declarations -fpermissive -IlnInclude -I. -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OpenFOAM/lnInclude -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OSspecific/POSIX/lnInclude   -fPIC -c PstreamGlobals.C -o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/PstreamGlobals.o
g++ -std=c++11 -m64 -DOPENFOAM=1906 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O2 -ftree-vectorize -march=native -fno-math-errno -std=c++11 -fuse-ld=bfd  -DNoRepository -ftemplate-depth-100   -Wno-old-style-cast -Wno-unused-local-typedefs -Wno-array-bounds -Wno-deprecated-declarations -fpermissive -IlnInclude -I. -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OpenFOAM/lnInclude -I/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/src/OSspecific/POSIX/lnInclude   -fPIC -shared -Xlinker --add-needed -Xlinker --no-as-needed /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UOPwrite.o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UIPread.o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/UPstream.o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/build/linux64GccDPInt32OptEASYBUILDMPI/src/Pstream/mpi/PstreamGlobals.o -L/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32OptEASYBUILDMPI/lib \
      -o /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/platforms/linux64GccDPInt32Opt/lib/mpi/libPstream.so
```
I have checked the rules directory and this is what I have.
```
$ grep -R "g++ -std=c++11" /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/General/Clang/c++:CC          = clang++ -std=c++11
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/General/Gcc/c++:CC          = g++ -std=c++11
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/linuxARM64Arm/c++.orig.eb:CC          = armclang++ -std=c++11
$ grep -R "mpicxx" /soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/linux64Cray/c++:CC          = OMPI_CXX="g++" mpicxx
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/linux64Mingw/c++:CC          = OMPI_CXX="g++" mpicxx
/soft/easybuild/software/OpenFOAM/v1906-foss-2018a/OpenFOAM-v1906/wmake/rules/linuxARM64Arm/c++:CC          = OMPI_CXX="g++" mpicxx
```

Any idea?

**Comment by mboisson** [comment]
@omula, did you take easybuilders/easybuild-easyblocks#1772 ?

**Comment by omula** [comment]
@mboisson sorry I was not aware of it.

Is there something similar to --from-pr for easyblocks?

Thanks

**Comment by mboisson** [comment]
I don't know, sorry.

**Comment by omula** [comment]
@mboisson already tested in and working. Many thanks

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
generoso-1 - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 v4 @ 2.00GHz, Python 3.6.8
See https://gist.github.com/1dadd747d374adf09f80eeb927e81d86 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
generoso-1 - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 v4 @ 2.00GHz, Python 3.6.8
See https://gist.github.com/a29f4a6e6245fdebc23bf8d1d569f95e for a full test report.

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ to solid dynamics and electromagnetics."""
+
+toolchain = {'name': 'foss', 'version': '2018b'}
+toolchainopts = {'cstd': 'c++11'}
+
+source_urls = ['https://sourceforge.net/projects/openfoam/files/%(version)s/']
+sources = [SOURCE_TGZ]
+patches = ['OpenFOAM-%(version)s-cleanup.patch']
+checksums = [
+    '15e38c2dc659b63753a0dd3dff913222cc46d6a40089a1b76973dd741145f61a',  # OpenFOAM-v1906.tgz
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb]: @mboisson Incorrect checksum?

Should be `bee03c4b1da0d2c9f98eb469eeffbce3a8614728ef8e87f664042a7490976537` (see also #9364, confirmed with a fresh download)

Comment (mboisson) [reply easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb]: Hem... that's the checksum that was current when I installed it on our systems...

Comment (mboisson) [reply easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb]: So, after investigation, OpenFOAM re-released v1906 apparently. Mostly changes in comments, but a few changes in the code. 

```
[mboisson@build-node ~]$ diff -ru OpenFOAM-v1906.{old,new} | grep -- '^[-+][^-+#]' | grep -v "Released\|Copyright\|www.openfoam.com\|A nd\|blob\|F ield\|O peration\|M anipulation\|======\|\/\/"
+%------------------------------------------------------------------------------
+
+  You can see how your report will be rendered on the platform by using the
+  "preview" tab above
-OpenFOAM version :
-Operating system :
-Compiler         :
+  The "/label ~bug" text is a gitlab flag that will add the "bug" label to this
+  issue
-head=641c8ae55e13b527ed99fa30a7d660dedee95dcc
+head=cebdbdff0e71ba850b2d771fe9747184d0de7961
-    return HashPtrTable<IOobject>::filterKeys(pred, pruning);
+    label changed = 0;
+    for (iterator iter = begin(); iter != end(); ++iter)
+    {
+        if
+        (
+            (pred(iter.key()) ? pruning : !pruning)
+         && erase(iter)
+        )
+        {
+            ++changed;
+        }
+    }
+    return changed;
-                    hpTa += sigma.value()*(emissivity_*lambdaTa4 + pow4(Ta));
+                    hpTa += emissivity_*sigma.value()*(lambdaTa4 + pow4(Ta));
-                    hpTa += sigma.value()*pow4(Ta);
+                    hpTa += emissivity_*sigma.value()*pow4(Ta);
-    if (debug)
-    {
-        const scalar Q = gSum(kappa(Tp)*patch().magSf()*snGrad());
-        Info<< patch().boundaryMesh().mesh().name() << ':'
-            << patch().name() << ':'
-            << internalField().name() << " :"
-            << " heat transfer rate:" << Q
-            << " wall temperature "
-            << " min:" << gMin(*this)
-            << " max:" << gMax(*this)
-            << " avg:" << gAverage(*this)
-            << endl;
-    }
+    DebugInfo
+        << patch().boundaryMesh().mesh().name() << ':' << patch().name() << ':'
+        << internalField().name() << " :"
+        << " heat transfer rate:" << gSum(kappa(Tp)*patch().magSf()*snGrad())
+        << " wall temperature "
+        << " min:" << gMin(*this)
+        << " max:" << gMax(*this)
+        << " avg:" << gAverage(*this) << nl;
```

**Comment by boegel** [comment]
@mboisson Is it worth pursuing this now #9364 is merged?

**Comment by mboisson** [comment]
I assume #9364 included the patches from this PR ? If so, then no, having a single one is just fine. We don't use foss anyway, so it is not directly applicable to Compute Canada, I opened the PR mostly to get the patches in.

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb (hunk 1 above)
Comment (boegel) [reply easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-v1906-foss-2018b.eb]: Sneaky bastards... They should be quartered.

**Comment by boegel** [comment]
Patch is indeed included in #9364, so closing this...

**Pull request closed without merging**